dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/spirv/VisitorOpsComposite.java

```diff
@@ -64,6 +66,53 @@ public Void visitOpCompositeInsert(SpirvParser.OpCompositeInsertContext ctx) {
         throw new ParsingException("Type mismatch in composite insert for '%s'", id);
     }
 
+    @Override
+    public Void visitOpVectorShuffle(SpirvParser.OpVectorShuffleContext ctx) {
+        String id = ctx.idResult().getText();
+        String v1Id = ctx.vector1().getText();
+        Expression v1 = builder.getExpression(v1Id);
+        String v2Id = ctx.vector2().getText();
+        Expression v2 = builder.getExpression(v2Id);
+        List<Integer> components = ctx.components().stream()
+            .map(SpirvParser.ComponentsContext::getText)
+            .map(Integer::parseInt)
+            .toList();
+        Type type = builder.getType(ctx.idResultType().getText());
+        if (type instanceof ArrayType aType) {
+            Type eType = aType.getElementType();
+            if (v1.getType() instanceof ArrayType aType1) {
+                if (!eType.equals(aType1.getElementType())) {
+                    throw new ParsingException("Type missmatch in OpVectorShuffle '%s' between result type %s and components %s", id, eType, aType1.getElementType());
+                }
```

Comment:
(1) The first two ifs can be inverted like the third one to avoid all this nesting:
```Java
if (!(type instance ArrayType aType)) {
   throw ...
}
if (!(v1.getType instance of ArrayType aType1)) {
   throw ...
}
```
This is also true for the `v2.getType` check afterwards. I think the main logic is then without any nesting levels.

(2) You can move `Type eType = aType.getElementType();` closer to its first usage (here: after the if).

(3) Also typo: "mismatch" is written with a single s.